File: apps/frontend/lib/models/tournament_models.dart
import 'package:json_annotation/json_annotation.dart';

part 'tournament_models.g.dart';

@JsonSerializable()
class LeaderboardEntry {
  final int rank;
  @JsonKey(name: 'user_id')
  final String userId;
  final String username;
  final double score;
  @JsonKey(name: 'is_ai', defaultValue: false)
  final bool isAi;
  final String? avatar;
  final String? league;
  @JsonKey(name: 'change_direction')
  final String? changeDirection;
  @JsonKey(name: 'change_amount')
  final int? changeAmount;

  LeaderboardEntry({
    required this.rank,
    required this.userId,
    required this.username,
    required this.score,
    required this.isAi,
    this.avatar,
    this.league,
    this.changeDirection,
    this.changeAmount,
  });

  factory LeaderboardEntry.fromJson(Map<String, dynamic> json) =>
      _$LeaderboardEntryFromJson(json);

  Map<String, dynamic> toJson() => _$LeaderboardEntryToJson(this);
}

@JsonSerializable()
class TournamentMetadata {
  @JsonKey(name: 'tournament_id')
  final String tournamentId;
  final String status;
  @JsonKey(name: 'start_time')
  final String startTime;
  @JsonKey(name: 'end_time')
  final String endTime;
  @JsonKey(name: 'total_participants')
  final int totalParticipants;
  @JsonKey(name: 'total_trades')
  final int totalTrades;
  @JsonKey(name: 'prize_pool')
  final double? prizePool;
  @JsonKey(name: 'entry_fee')
  final double? entryFee;

  TournamentMetadata({
    required this.tournamentId,
    required this.status,
    required this.startTime,
    required this.endTime,
    required this.totalParticipants,
    required this.totalTrades,
    this.prizePool,
    this.entryFee,
  });

  factory TournamentMetadata.fromJson(Map<String, dynamic> json) =>
      _$TournamentMetadataFromJson(json);

  Map<String, dynamic> toJson() => _$TournamentMetadataToJson(this);
}

@JsonSerializable()
class UserTournamentStats {
  @JsonKey(name: 'user_id')
  final String userId;
  @JsonKey(name: 'total_trades')
  final int totalTrades;
  @JsonKey(name: 'winning_trades')
  final int winningTrades;
  @JsonKey(name: 'total_profit')
  final double totalProfit;
  @JsonKey(name: 'total_volume')
  final double totalVolume;
  @JsonKey(name: 'win_rate')
  final double winRate;
  @JsonKey(name: 'portfolio_value')
  final double portfolioValue;
  final int? rank;
  final double? percentile;

  UserTournamentStats({
    required this.userId,
    required this.totalTrades,
    required this.winningTrades,
    required this.totalProfit,
    required this.totalVolume,
    required this.winRate,
    required this.portfolioValue,
    this.rank,
    this.percentile,
  });

  factory UserTournamentStats.fromJson(Map<String, dynamic> json) =>
      _$UserTournamentStatsFromJson(json);

  Map<String, dynamic> toJson() => _$UserTournamentStatsToJson(this);
}

@JsonSerializable()
class LiveTrade {
  @JsonKey(name: 'trade_id')
  final String tradeId;
  @JsonKey(name: 'user_id')
  final String userId;
  final String username;
  final String symbol;
  final String side;
  final double amount;
  final double? price;
  final double? profit;
  @JsonKey(name: 'profit_percentage')
  final double? profitPercentage;
  @JsonKey(name: 'is_ai', defaultValue: false)
  final bool isAi;
  @JsonKey(name: 'ai_strategy')
  final String? aiStrategy;
  final String timestamp;

  LiveTrade({
    required this.tradeId,
    required this.userId,
    required this.username,
    required this.symbol,
    required this.side,
    required this.amount,
    this.price,
    this.profit,
    this.profitPercentage,
    required this.isAi,
    this.aiStrategy,
    required this.timestamp,
  });

  factory LiveTrade.fromJson(Map<String, dynamic> json) =>
      _$LiveTradeFromJson(json);

  Map<String, dynamic> toJson() => _$LiveTradeToJson(this);
}

@JsonSerializable()
class Achievement {
  @JsonKey(name: 'achievement_id')
  final String achievementId;
  @JsonKey(name: 'user_id')
  final String userId;
  final String username;
  final String title;
  final String description;
  final String icon;
  final int points;
  final String rarity;
  final String category;
  @JsonKey(name: 'unlocked_at')
  final String unlockedAt;

  Achievement({
    required this.achievementId,
    required this.userId,
    required this.username,
    required this.title,
    required this.description,
    required this.icon,
    required this.points,
    required this.rarity,
    required this.category,
    required this.unlockedAt,
  });

  factory Achievement.fromJson(Map<String, dynamic> json) =>
      _$AchievementFromJson(json);

  Map<String, dynamic> toJson() => _$AchievementToJson(this);
}

@JsonSerializable()
class AITraderInfo {
  final String id;
  final String name;
  final String title;
  final String strategy;
  final String backstory;
  @JsonKey(name: 'risk_profile')
  final Map<String, double> riskProfile;
  @JsonKey(name: 'personality_traits')
  final Map<String, double> personalityTraits;
  @JsonKey(name: 'preferred_symbols')
  final List<String> preferredSymbols;
  @JsonKey(name: 'trading_hours')
  final List<int> tradingHours;
  @JsonKey(name: 'current_state')
  final Map<String, dynamic> currentState;
  @JsonKey(name: 'performance_stats')
  final Map<String, dynamic>? performanceStats;

  AITraderInfo({
    required this.id,
    required this.name,
    required this.title,
    required this.strategy,
    required this.backstory,
    required this.riskProfile,
    required this.personalityTraits,
    required this.preferredSymbols,
    required this.tradingHours,
    required this.currentState,
    this.performanceStats,
  });

  factory AITraderInfo.fromJson(Map<String, dynamic> json) =>
      _$AITraderInfoFromJson(json);

  Map<String, dynamic> toJson() => _$AITraderInfoToJson(this);
}

@JsonSerializable()
class TournamentEvent {
  @JsonKey(name: 'event_id')
  final String eventId;
  @JsonKey(name: 'event_type')
  final String eventType;
  final String title;
  final String description;
  @JsonKey(name: 'duration_minutes')
  final int? durationMinutes;
  final double? multiplier;
  @JsonKey(name: 'affects_leaderboard')
  final bool affectsLeaderboard;
  @JsonKey(name: 'started_at')
  final String startedAt;
  @JsonKey(name: 'ends_at')
  final String? endsAt;

  TournamentEvent({
    required this.eventId,
    required this.eventType,
    required this.title,
    required this.description,
    this.durationMinutes,
    this.multiplier,
    required this.affectsLeaderboard,
    required this.startedAt,
    this.endsAt,
  });

  factory TournamentEvent.fromJson(Map<String, dynamic> json) =>
      _$TournamentEventFromJson(json);

  Map<String, dynamic> toJson() => _$TournamentEventToJson(this);
}

@JsonSerializable()
class InitialStateData {
  @JsonKey(name: 'tournament_id')
  final String tournamentId;
  final List<LeaderboardEntry> leaderboard;
  @JsonKey(name: 'user_rank')
  final int? userRank;
  @JsonKey(name: 'user_score')
  final double? userScore;
  @JsonKey(name: 'user_stats')
  final UserTournamentStats? userStats;
  @JsonKey(name: 'active_connections')
  final int activeConnections;
  @JsonKey(name: 'tournament_meta')
  final TournamentMetadata tournamentMeta;
  @JsonKey(name: 'server_time')
  final double serverTime;
  @JsonKey(name: 'ai_traders_count')
  final int aiTradersCount;
  @JsonKey(name: 'recent_trades', defaultValue: <LiveTrade>[])
  final List<LiveTrade> recentTrades;

  InitialStateData({
    required this.tournamentId,
    required this.leaderboard,
    this.userRank,
    this.userScore,
    this.userStats,
    required this.activeConnections,
    required this.tournamentMeta,
    required this.serverTime,
    required this.aiTradersCount,
    required this.recentTrades,
  });

  factory InitialStateData.fromJson(Map<String, dynamic> json) =>
      _$InitialStateDataFromJson(json);

  Map<String, dynamic> toJson() => _$InitialStateDataToJson(this);
}

@JsonSerializable()
class LeaderboardUpdateData {
  @JsonKey(name: 'user_id')
  final String userId;
  final String username;
  @JsonKey(name: 'old_rank')
  final int? oldRank;
  @JsonKey(name: 'new_rank')
  final int newRank;
  @JsonKey(name: 'old_score')
  final double? oldScore;
  @JsonKey(name: 'new_score')
  final double newScore;
  @JsonKey(name: 'rank_change', defaultValue: 0)
  final int rankChange;
  @JsonKey(name: 'score_change', defaultValue: 0.0)
  final double scoreChange;
  @JsonKey(name: 'is_ai', defaultValue: false)
  final bool isAi;

  LeaderboardUpdateData({
    required this.userId,
    required this.username,
    this.oldRank,
    required this.newRank,
    this.oldScore,
    required this.newScore,
    required this.rankChange,
    required this.scoreChange,
    required this.isAi,
  });

  factory LeaderboardUpdateData.fromJson(Map<String, dynamic> json) =>
      _$LeaderboardUpdateDataFromJson(json);

  Map<String, dynamic> toJson() => _$LeaderboardUpdateDataToJson(this);
}

@JsonSerializable()
class NearbyCompetitorsData {
  final List<LeaderboardEntry> competitors;
  @JsonKey(name: 'user_rank')
  final int userRank;
  final int radius;
  @JsonKey(name: 'total_participants')
  final int totalParticipants;

  NearbyCompetitorsData({
    required this.competitors,
    required this.userRank,
    required this.radius,
    required this.totalParticipants,
  });

  factory NearbyCompetitorsData.fromJson(Map<String, dynamic> json) =>
      _$NearbyCompetitorsDataFromJson(json);

  Map<String, dynamic> toJson() => _$NearbyCompetitorsDataToJson(this);
}

@JsonSerializable()
class AuthenticationData {
  @JsonKey(name: 'user_id')
  final String userId;
  final String? username;
  final bool authenticated;
  final List<String> permissions;

  AuthenticationData({
    required this.userId,
    this.username,
    required this.authenticated,
    required this.permissions,
  });

  factory AuthenticationData.fromJson(Map<String, dynamic> json) =>
      _$AuthenticationDataFromJson(json);

  Map<String, dynamic> toJson() => _$AuthenticationDataToJson(this);
}

@JsonSerializable()
class ErrorData {
  @JsonKey(name: 'error_code')
  final String errorCode;
  @JsonKey(name: 'error_message')
  final String errorMessage;
  final Map<String, dynamic>? details;
  @JsonKey(name: 'retry_after')
  final int? retryAfter;

  ErrorData({
    required this.errorCode,
    required this.errorMessage,
    this.details,
    this.retryAfter,
  });

  factory ErrorData.fromJson(Map<String, dynamic> json) =>
      _$ErrorDataFromJson(json);

  Map<String, dynamic> toJson() => _$ErrorDataToJson(this);
}

@JsonSerializable()
class ConnectionStats {
  @JsonKey(name: 'total_connections')
  final int totalConnections;
  @JsonKey(name: 'peak_concurrent')
  final int peakConcurrent;
  @JsonKey(name: 'messages_sent')
  final int messagesSent;
  @JsonKey(name: 'messages_received')
  final int messagesReceived;
  final int reconnections;
  @JsonKey(name: 'current_connections')
  final int currentConnections;
  @JsonKey(name: 'active_tournaments')
  final int activeTournaments;
  @JsonKey(name: 'tournament_breakdown')
  final Map<String, int> tournamentBreakdown;

  ConnectionStats({
    required this.totalConnections,
    required this.peakConcurrent,
    required this.messagesSent,
    required this.messagesReceived,
    required this.reconnections,
    required this.currentConnections,
    required this.activeTournaments,
    required this.tournamentBreakdown,
  });

  factory ConnectionStats.fromJson(Map<String, dynamic> json) =>
      _$ConnectionStatsFromJson(json);

  Map<String, dynamic> toJson() => _$ConnectionStatsToJson(this);
}

// Utility classes for UI state management
class TournamentUIState {
  final List<LeaderboardEntry> leaderboard;
  final List<LiveTrade> recentTrades;
  final List<Achievement> recentAchievements;
  final TournamentEvent? activeEvent;
  final UserTournamentStats? userStats;
  final int? userRank;
  final double? userScore;
  final bool isConnected;
  final String? errorMessage;
  final DateTime lastUpdate;

  TournamentUIState({
    this.leaderboard = const [],
    this.recentTrades = const [],
    this.recentAchievements = const [],
    this.activeEvent,
    this.userStats,
    this.userRank,
    this.userScore,
    this.isConnected = false,
    this.errorMessage,
    DateTime? lastUpdate,
  }) : lastUpdate = lastUpdate ?? DateTime.now();

  TournamentUIState copyWith({
    List<LeaderboardEntry>? leaderboard,
    List<LiveTrade>? recentTrades,
    List<Achievement>? recentAchievements,
    TournamentEvent? activeEvent,
    UserTournamentStats? userStats,
    int? userRank,
    double? userScore,
    bool? isConnected,
    String? errorMessage,
    DateTime? lastUpdate,
  }) {
    return TournamentUIState(
      leaderboard: leaderboard ?? this.leaderboard,
      recentTrades: recentTrades ?? this.recentTrades,
      recentAchievements: recentAchievements ?? this.recentAchievements,
      activeEvent: activeEvent ?? this.activeEvent,
      userStats: userStats ?? this.userStats,
      userRank: userRank ?? this.userRank,
      userScore: userScore ?? this.userScore,
      isConnected: isConnected ?? this.isConnected,
      errorMessage: errorMessage,
      lastUpdate: lastUpdate ?? DateTime.now(),
    );
  }
}

// Tournament notification types
enum TournamentNotificationType {
  leaderboardChange,
  achievementUnlocked,
  tournamentEvent,
  aiTradeAlert,
  competitorNearby,
  personalBest,
}

class TournamentNotification {
  final TournamentNotificationType type;
  final String title;
  final String message;
  final String? imageUrl;
  final Map<String, dynamic>? data;
  final DateTime timestamp;
  final Duration? displayDuration;

  TournamentNotification({
    required this.type,
    required this.title,
    required this.message,
    this.imageUrl,
    this.data,
    DateTime? timestamp,
    this.displayDuration,
  }) : timestamp = timestamp ?? DateTime.now();
}

// WebSocket message types enum
enum WebSocketMessageType {
  initialState,
  leaderboardUpdate,
  liveTrade,
  achievementUnlocked,
  tournamentEvent,
  heartbeat,
  nearbyCompetitors,
  aiTraderInfo,
  detailedLeaderboard,
  authenticationSuccess,
  error,
  getNearbyCompetitors,
  getAITraderInfo,
  getDetailedLeaderboard,
  authenticate,
  heartbeatResponse,
}

File: apps/backend/services/competition/tournament_models.py
from typing import Dict, List, Optional, Any, Union
from datetime import datetime
from pydantic import BaseModel, Field
from enum import Enum

class WebSocketMessageType(Enum):
    """WebSocket message types"""
    # Server to client messages
    INITIAL_STATE = "initial_state"
    LEADERBOARD_UPDATE = "leaderboard_update"
    LIVE_TRADE = "live_trade"
    ACHIEVEMENT_UNLOCKED = "achievement_unlocked"
    TOURNAMENT_EVENT = "tournament_event"
    HEARTBEAT = "heartbeat"
    NEARBY_COMPETITORS = "nearby_competitors"
    AI_TRADER_INFO = "ai_trader_info"
    DETAILED_LEADERBOARD = "detailed_leaderboard"
    AUTHENTICATION_SUCCESS = "authentication_success"
    ERROR = "error"
    
    # Client to server messages
    GET_NEARBY_COMPETITORS = "get_nearby_competitors"
    GET_AI_TRADER_INFO = "get_ai_trader_info"
    GET_DETAILED_LEADERBOARD = "get_detailed_leaderboard"
    AUTHENTICATE = "authenticate"
    HEARTBEAT_RESPONSE = "heartbeat_response"

class ConnectionState(Enum):
    """WebSocket connection states"""
    CONNECTING = "connecting"
    CONNECTED = "connected"
    AUTHENTICATED = "authenticated"
    DISCONNECTED = "disconnected"

# Base message models
class BaseWebSocketMessage(BaseModel):
    """Base WebSocket message structure"""
    type: str
    timestamp: str = Field(default_factory=lambda: datetime.utcnow().isoformat())
    
class ClientMessage(BaseWebSocketMessage):
    """Client to server message"""
    data: Optional[Dict[str, Any]] = None

class ServerMessage(BaseWebSocketMessage):
    """Server to client message"""
    data: Optional[Dict[str, Any]] = None
    error: Optional[str] = None

# Tournament data models
class LeaderboardEntry(BaseModel):
    """Tournament leaderboard entry"""
    rank: int
    user_id: str
    username: str
    score: float
    is_ai: bool = False
    avatar: Optional[str] = None
    league: Optional[str] = None
    change_direction: Optional[str] = None  # "up", "down", "same"
    change_amount: Optional[int] = None

class TournamentMetadata(BaseModel):
    """Tournament metadata"""
    tournament_id: str
    status: str  # "active", "ended", "upcoming"
    start_time: str
    end_time: str
    total_participants: int
    total_trades: int
    prize_pool: Optional[float] = None
    entry_fee: Optional[float] = None

class UserTournamentStats(BaseModel):
    """User's tournament statistics"""
    user_id: str
    total_trades: int
    winning_trades: int
    total_profit: float
    total_volume: float
    win_rate: float
    portfolio_value: float
    rank: Optional[int] = None
    percentile: Optional[float] = None

class LiveTrade(BaseModel):
    """Live trade notification"""
    trade_id: str
    user_id: str
    username: str
    symbol: str
    side: str  # "BUY" or "SELL"
    amount: float
    price: Optional[float] = None
    profit: Optional[float] = None
    profit_percentage: Optional[float] = None
    is_ai: bool = False
    ai_strategy: Optional[str] = None
    timestamp: str

class Achievement(BaseModel):
    """Achievement unlocked"""
    achievement_id: str
    user_id: str
    username: str
    title: str
    description: str
    icon: str
    points: int
    rarity: str  # "common", "rare", "epic", "legendary"
    category: str  # "trading", "social", "progression", "special"
    unlocked_at: str

class AITraderInfo(BaseModel):
    """AI Trader detailed information"""
    id: str
    name: str
    title: str
    strategy: str
    backstory: str
    risk_profile: Dict[str, float]
    personality_traits: Dict[str, float]
    preferred_symbols: List[str]
    trading_hours: List[int]
    current_state: Dict[str, Any]
    performance_stats: Optional[Dict[str, Any]] = None

class TournamentEvent(BaseModel):
    """Special tournament events"""
    event_id: str
    event_type: str  # "bonus_round", "market_crash", "volatility_spike", "special_announcement"
    title: str
    description: str
    duration_minutes: Optional[int] = None
    multiplier: Optional[float] = None
    affects_leaderboard: bool = True
    started_at: str
    ends_at: Optional[str] = None

# Server to Client message data models
class InitialStateData(BaseModel):
    """Initial state data sent on connection"""
    tournament_id: str
    leaderboard: List[LeaderboardEntry]
    user_rank: Optional[int] = None
    user_score: Optional[float] = None
    user_stats: Optional[UserTournamentStats] = None
    active_connections: int
    tournament_meta: TournamentMetadata
    server_time: float
    ai_traders_count: int
    recent_trades: List[LiveTrade] = []

class LeaderboardUpdateData(BaseModel):
    """Leaderboard position change update"""
    user_id: str
    username: str
    old_rank: Optional[int] = None
    new_rank: int
    old_score: Optional[float] = None
    new_score: float
    rank_change: int = 0  # Positive = moved up, negative = moved down
    score_change: float = 0.0
    is_ai: bool = False

class NearbyCompetitorsData(BaseModel):
    """Nearby competitors response"""
    competitors: List[LeaderboardEntry]
    user_rank: int
    radius: int
    total_participants: int

class DetailedLeaderboardData(BaseModel):
    """Detailed leaderboard with stats"""
    leaderboard: List[Dict[str, Any]]  # Enhanced leaderboard entries with stats
    start: int
    limit: int
    total_participants: int
    user_rank: Optional[int] = None

class AuthenticationData(BaseModel):
    """Authentication success response"""
    user_id: str
    username: Optional[str] = None
    authenticated: bool = True
    permissions: List[str] = []

class HeartbeatData(BaseModel):
    """Heartbeat message data"""
    server_time: float
    connection_id: Optional[str] = None
    uptime_seconds: Optional[float] = None

class ErrorData(BaseModel):
    """Error message data"""
    error_code: str
    error_message: str
    details: Optional[Dict[str, Any]] = None
    retry_after: Optional[int] = None

# Client to Server message data models
class GetNearbyCompetitorsRequest(BaseModel):
    """Request nearby competitors"""
    radius: int = Field(default=5, ge=1, le=20)
    
class GetAITraderInfoRequest(BaseModel):
    """Request AI trader information"""
    ai_id: str

class GetDetailedLeaderboardRequest(BaseModel):
    """Request detailed leaderboard"""
    start: int = Field(default=0, ge=0)
    limit: int = Field(default=50, ge=1, le=100)

class AuthenticateRequest(BaseModel):
    """Authentication request"""
    token: str
    refresh_token: Optional[str] = None

class HeartbeatResponseRequest(BaseModel):
    """Heartbeat response from client"""
    received_at: float
    client_time: float

# Complete message type definitions
class InitialStateMessage(ServerMessage):
    """Initial state message"""
    type: str = WebSocketMessageType.INITIAL_STATE.value
    data: InitialStateData

class LeaderboardUpdateMessage(ServerMessage):
    """Leaderboard update message"""
    type: str = WebSocketMessageType.LEADERBOARD_UPDATE.value
    data: LeaderboardUpdateData

class LiveTradeMessage(ServerMessage):
    """Live trade message"""
    type: str = WebSocketMessageType.LIVE_TRADE.value
    data: LiveTrade

class AchievementUnlockedMessage(ServerMessage):
    """Achievement unlocked message"""
    type: str = WebSocketMessageType.ACHIEVEMENT_UNLOCKED.value
    data: Achievement

class TournamentEventMessage(ServerMessage):
    """Tournament event message"""
    type: str = WebSocketMessageType.TOURNAMENT_EVENT.value
    data: TournamentEvent

class HeartbeatMessage(ServerMessage):
    """Heartbeat message"""
    type: str = WebSocketMessageType.HEARTBEAT.value
    data: HeartbeatData

class NearbyCompetitorsMessage(ServerMessage):
    """Nearby competitors message"""
    type: str = WebSocketMessageType.NEARBY_COMPETITORS.value
    data: NearbyCompetitorsData

class AITraderInfoMessage(ServerMessage):
    """AI trader info message"""
    type: str = WebSocketMessageType.AI_TRADER_INFO.value
    data: AITraderInfo

class DetailedLeaderboardMessage(ServerMessage):
    """Detailed leaderboard message"""
    type: str = WebSocketMessageType.DETAILED_LEADERBOARD.value
    data: DetailedLeaderboardData

class AuthenticationSuccessMessage(ServerMessage):
    """Authentication success message"""
    type: str = WebSocketMessageType.AUTHENTICATION_SUCCESS.value
    data: AuthenticationData

class ErrorMessage(ServerMessage):
    """Error message"""
    type: str = WebSocketMessageType.ERROR.value
    data: ErrorData

# Message parsing and validation
class MessageParser:
    """Utility class for parsing and validating WebSocket messages"""
    
    @staticmethod
    def parse_client_message(data: Dict[str, Any]) -> Optional[BaseModel]:
        """Parse incoming client message"""
        try:
            msg_type = data.get('type')
            
            if msg_type == WebSocketMessageType.GET_NEARBY_COMPETITORS.value:
                return GetNearbyCompetitorsRequest(**data.get('data', {}))
            elif msg_type == WebSocketMessageType.GET_AI_TRADER_INFO.value:
                return GetAITraderInfoRequest(**data.get('data', {}))
            elif msg_type == WebSocketMessageType.GET_DETAILED_LEADERBOARD.value:
                return GetDetailedLeaderboardRequest(**data.get('data', {}))
            elif msg_type == WebSocketMessageType.AUTHENTICATE.value:
                return AuthenticateRequest(**data.get('data', {}))
            elif msg_type == WebSocketMessageType.HEARTBEAT_RESPONSE.value:
                return HeartbeatResponseRequest(**data.get('data', {}))
            
            return None
        except Exception:
            return None
    
    @staticmethod
    def create_server_message(msg_type: WebSocketMessageType, data: Any, error: str = None) -> Dict[str, Any]:
        """Create server message dict"""
        message = {
            'type': msg_type.value,
            'timestamp': datetime.utcnow().isoformat()
        }
        
        if error:
            message['error'] = error
        else:
            message['data'] = data.dict() if hasattr(data, 'dict') else data
            
        return message

# Connection statistics models
class ConnectionStats(BaseModel):
    """WebSocket connection statistics"""
    total_connections: int
    peak_concurrent: int
    messages_sent: int
    messages_received: int
    reconnections: int
    current_connections: int
    active_tournaments: int
    tournament_breakdown: Dict[str, int]

class TournamentConnectionInfo(BaseModel):
    """Tournament connection information"""
    tournament_id: str
    active_connections: int
    authenticated_users: int
    anonymous_users: int
    peak_connections: int
    total_messages_sent: int
    total_messages_received: int
    average_session_duration: Optional[float] = None

# Rate limiting models
class RateLimitInfo(BaseModel):
    """Rate limiting information"""
    limit_per_minute: int
    current_usage: int
    reset_time: str
    blocked_until: Optional[str] = None

# Health check models
class WebSocketHealthCheck(BaseModel):
    """WebSocket health check response"""
    status: str  # "healthy", "degraded", "unhealthy"
    active_connections: int
    active_tournaments: int
    redis_connected: bool
    pubsub_channels: int
    last_heartbeat: str
    uptime_seconds: float
    memory_usage_mb: Optional[float] = None
    cpu_usage_percent: Optional[float] = None

# Tournament phase models
class TournamentPhase(Enum):
    """Tournament phases"""
    REGISTRATION = "registration"
    STARTING = "starting"
    ACTIVE = "active"
    CLOSING = "closing"
    ENDED = "ended"
    ARCHIVED = "archived"

class TournamentStatusUpdate(BaseModel):
    """Tournament status update"""
    tournament_id: str
    old_phase: TournamentPhase
    new_phase: TournamentPhase
    phase_start_time: str
    phase_end_time: Optional[str] = None
    participant_count: int
    total_trades: int
    message: Optional[str] = None

# Leaderboard snapshot models
class LeaderboardSnapshot(BaseModel):
    """Leaderboard snapshot for historical data"""
    tournament_id: str
    snapshot_time: str
    top_performers: List[LeaderboardEntry]
    total_participants: int
    snapshot_type: str  # "hourly", "daily", "final"
    
# Market condition models (for AI trading context)
class MarketConditionsUpdate(BaseModel):
    """Market conditions update for AI trading"""
    timestamp: str
    volatility: float
    trend_direction: str  # "UP", "DOWN", "SIDEWAYS"
    trend_strength: float
    volume_profile: Dict[str, Any]
    sentiment_score: float
    major_events: List[str] = []

# Event aggregation models
class EventSummary(BaseModel):
    """Summary of events over time period"""
    time_period: str
    total_trades: int
    unique_traders: int
    ai_trades: int
    human_trades: int
    top_performers: List[str]
    major_movements: List[LeaderboardUpdateData]
    achievements_unlocked: int
    tournament_events: List[TournamentEvent]